text/0000-guaranteed-non-static-destructors.md

```diff
@@ -0,0 +1,180 @@
+ensure that all stack-anchored objects are destructed. Rust is primarily
+concerned with memory safety, so this RFC tries to formulate a general rule
+that addresses the issue of incorrect assumptions about destructors leading to
+memory unsafety. It does not attempt to prevent the leaking of other resources,
+as leaking a `'static` object is never unsafe. (`'static` objects are
+conceptually safe to keep around “forever”, so any unsoundness that can be
+obtained by a leak could also be obtained through other means).
+
+# Detailed design
+
+In addition to current guarantees, the following property can be relied upon:
+
+Given an object `A` implementing `Drop` whose lifetime is restricted to `'a`
+(that is, the object may not live longer than `'a`), in absence of unsafe code,
+the destructor for `A` is guaranteed to run before any code after the end of
+`'a` is executed.
+
+This is already true at the language level. However, certain library constructs
+currently allow safe code to break it.
+
+This has the following implications:
+
+ * It is perfectly acceptable to forget, leak, et cetera any object as long as
+   that object is valid for the duration of `'static`. Intuitively, this makes
+   sense, as forgetting an object is the same as having it live forever.
+ * A resource that is never freed due to and endless loop, a deadlock, or
+   program termination is valid because any code that comes after the end of
+   `'a` is never executed.
+ * Code would be allowed to rely on the guarantee for soundness. This means
+   that patterns such as `thread::scoped` and the initially-proposed
+   `Vec::drain_range` would be sound.
+ * Unsafe code is free to forget objects as needed in cases where the
+   programmer guarantees it is sound. However, it is not allowed to expose a
+   safe interface that would allow the above guarantee to be violated.
+
+As noted, this guarantee is already true at the core language level. However,
+it can be violated in safe code when using the standard library. There are
+two known ways in which this can happen: when a destructor panics in certain
+situations (e.g., in a `Vec`), and when a reference cycle is created using `Rc`
+and friends.
+
+This RFC proposes the following solutions:
+
+ * Specify that any panic that occurs while a destructor is in progress results
+   in an abort. Panicking in a destructor is generally a bad idea with many
+   edge cases, so this is probably desirable, anyway. It should be possible to
+   implement this efficiently in a similar manner to C++’s `noexcept`.
+ * Restrict the basic `new` operation of existing reference-counted types to
+   types valid for `'static`, which are always safe to leak. Additionally,
+   introduce a scoped cycle collector that can be used to safely create `Rc`s
+   with a shorter lifetime, and research the possibility of a reference-counted
+   type that statically disallows cycles.
+
+Specifically, the scoped cycle collector would operate as follows:
```

Comment:
It would really help to have an implementation, or at least the signatures and types of all the parts of this API. It's quite difficult to judge the ergonomics and such of this proposal without them.